lib/trino-filesystem-alluxio/src/main/java/io/trino/filesystem/alluxio/AlluxioInputStream.java

```diff
@@ -0,0 +1,200 @@
+            throw new IOException("Output stream closed: " + location);
+        }
+    }
+
+    @Override
+    public int read()
+            throws IOException
+    {
+        ensureOpen();
+
+        byte[] bytes = new byte[1];
+        int n = read(bytes, 0, 1);
+        if (n == 1) {
+            return bytes[0] & 0xff;
+        }
+        if (n == -1) {
+            return -1;
+        }
+        throw new RuntimeException("Impossible");
+    }
+
+    @Override
+    public int read(byte[] bytes, int offset, int length)
+            throws IOException
+    {
+        ensureOpen();
+
+        checkFromIndexSize(offset, length, bytes.length);
+        if (position >= fileLength) {
+            return -1;
+        }
+        int bytesRead = doRead(bytes, offset, length);
+        position += bytesRead;
+        return bytesRead;
+    }
+
+    private int doRead(byte[] bytes, int offset, int length)
+            throws IOException
+    {
+        int bytesRead = 0;
+        bytesRead += helper.doBufferRead(position + bytesRead, bytes, offset + bytesRead, length - bytesRead);
+        bytesRead = addExact(bytesRead, helper.doCacheRead(position + bytesRead, bytes, offset + bytesRead, length - bytesRead));
+        bytesRead = addExact(bytesRead, doExternalRead(position + bytesRead, bytes, offset + bytesRead, length - bytesRead));
+        return bytesRead;
+    }
+
+    private int doExternalRead(long readPosition, byte[] buffer, int offset, int length)
+            throws IOException
+    {
+        if (length == 0) {
+            return 0;
+        }
+        AlluxioInputHelper.PageAlignedRead aligned = helper.alignRead(readPosition, length);
+        if (externalStream == null) {
+            externalStream = inputFile.newStream();
+        }
+        externalStream.seek(aligned.pageStart());
+        byte[] readBuffer = new byte[aligned.length()];
+        int externalBytesRead = externalStream.read(readBuffer, 0, aligned.length());
+        if (externalBytesRead < 0) {
+            throw new IOException("Unexpected end of stream");
+        }
+        helper.putCache(aligned.pageStart(), aligned.pageEnd(), readBuffer, externalBytesRead);
+        int bytesToCopy = min(length, max(externalBytesRead - aligned.pageOffset(), 0));
+        System.arraycopy(readBuffer, aligned.pageOffset(), buffer, offset, bytesToCopy);
+        statistics.recordExternalRead(inputFile.location(), externalBytesRead);
+        return bytesToCopy;
+    }
+
+    @Override
+    public long skip(long n)
+            throws IOException
+    {
+        ensureOpen();
+
+        n = Longs.constrainToRange(n, 0, fileLength - position);
+        position += n;
+        return n;
+    }
+
+    @Override
+    public void skipNBytes(long n)
+            throws IOException
+    {
+        ensureOpen();
+
+        if (n <= 0) {
+            return;
+        }
+
+        long position = this.position + n;
+        if ((position < 0) || (position > fileLength)) {
```

Comment:
if n > 0 always holds and  position is in the [0, Integer.MAX_VALUE] range, how it can be negative?

Reply:
It should protect against `this.position + n` overflowing. We have the same code for the other file system implementations
https://github.com/trinodb/trino/blob/1668ef349ad224b24ae4f40a7ec1401dc6cf2252/lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3InputStream.java#L143

Reply:
Use Math.addExact to guard against overflows

Reply:
Do we really want it? This would make the code look like
```java
        long position;
        try {
             position = Math.addExact(this.position, n);
        } catch (ArithmeticException e) {
            throw new EOFException("Unable to skip %s bytes (position=%s, fileSize=%s): %s".formatted(n, this.position, fileLength, location));
        }
        if ((position > fileLength)) {
            throw new EOFException("Unable to skip %s bytes (position=%s, fileSize=%s): %s".formatted(n, this.position, fileLength, location));
        }
```
and diverge from the other file system implementations.